File: 2020/day01-2.ts
function solution(content: string) {
  return content
    .split("\n")
    .map((item) => parseInt(item))
    .toSorted()
    .map((e, index, array) => {
      const found = find(2020 - e, [
        ...array.slice(0, index),
        ...array.slice(index + 1),
      ]);
      if (found !== null) return found * e;
      return null;
    })
    .filter((item) => item !== null)
    .at(0);
}

function find(target: number, list: number[]) {
  if (list.length < 2) return null;
  const fst = list.at(0)!;
  const lst = list.at(-1)!;
  const result = fst + lst;
  if (result > target) return find(target, list.slice(0, -1));
  if (result < target) return find(target, list.slice(1));
  return fst * lst;
}

Deno.readTextFile("day01.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2020/day01_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    expenses = sorted(int(x) for x in content if x)
    for index, e in enumerate(expenses):
        result = find(2020 - e, expenses[:index] + expenses[index + 1 :])
        if result is not None:
            return result * e


def find(target: int, expenses: list[int]):
    if len(expenses) < 2:
        return None
    fst, *rest, lst = expenses
    result = fst + lst
    if result > target:
        return find(target, [fst] + rest)
    if result < target:
        return find(target, rest + [lst])
    return fst * lst


def main() -> None:
    with open("day01.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
